### PR TITLE
Kernel Telemetry

### PR DESCRIPTION
**Pull request by alheinecke**
This is first PR to discuss the proposed changes

- it fixes a compilation bug in edge-v
- it adds per thread based cycle count of LIBXSMM fused JIT kernels
- it had to declare integral function non const
- exec_mm_kernel could be also a macro

### CODE REVIEW

File: src/data/MmXsmmFused.hpp [hunk 1]
```diff
@@ -180,6 +223,14 @@ class edge::data::MmXsmmFused< float > {
...
         m_descs.resize( i_group+1 );
         m_kernels.resize( i_group+1 );
+        m_kernel_flops.resize( i_group+1 );
+#ifdef _OPENMP
+        for ( int i = 0; i < omp_get_max_threads(); ++i ) {
+          m_kernel_stats[i].resize( i_group+1 );
+        }
+#else
+        m_kernel_stats[0].resize( i_group+1 );
+#endif
```
Comment (breuera) [inline src/data/MmXsmmFused.hpp]: Suggestion: Lets put the ``omp_get_max_threads()`` into the global vars and get rid of the preprocessor stuff, similar to https://github.com/3343/edge/blob/develop/src/parallel/global.h#L49

File: src/data/MmXsmmFused.hpp [hunk 2]
```diff
@@ -211,11 +262,26 @@ class edge::data::MmXsmmFused< float > {
         delete[] l_rows; delete[] l_cols; delete[] l_vals;
       }
       else {
-        m_kernels[i_group].push_back( libxsmm_create_rm_ac_soa( m_descs[i_group].back() ).smm );
+        m_kernels[i_group].push_back( libxsmm_create_pgemm_ac_rm( m_descs[i_group].back(), PP_N_CRUNS ).smm );
```
Comment (breuera) [inline src/data/MmXsmmFused.hpp]: There's a non-preprocessor version of `PP_N_CRUNS` called `N_CRUNS`: https://github.com/3343/edge/blob/develop/src/constants.hpp#L285

File: src/impl/seismic/kernels/SurfIntFused.hpp [hunk 3]
```diff
@@ -31,6 +31,12 @@
 #include "data/MmXsmmFused.hpp"
 #include "FakeMats.hpp"
 
+#ifdef PP_MMKERNEL_PERF
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+#endif
```
Comment (breuera) [inline src/impl/seismic/kernels/SurfIntFused.hpp]: Probably another argument for a global var so that we don't have to include OMP all over the place.

File: src/impl/seismic/kernels/SurfIntFused.hpp [hunk 4]
```diff
@@ -382,6 +388,51 @@ class edge::seismic::kernels::SurfIntFused: public edge::seismic::kernels::SurfI
...
+    }
+
+    inline void exec_mm_kernel_pf( const unsigned short    i_group, 
+                                  const unsigned short     i_kernel,
+                                  const TL_T_REAL *a,
+                                  const TL_T_REAL *b,
+                                        TL_T_REAL *c,
+                                  const TL_T_REAL *a_pf,
+                                  const TL_T_REAL *b_pf,
+                                  const TL_T_REAL *c_pf  ) {
```
Comment (breuera) [inline src/impl/seismic/kernels/SurfIntFused.hpp]: Add respective underscores `i_a`, [...], `io_c`; fix indentation, rename function (no underscores).
Wouldn't it be better to move this wrapper to the kernels themselves?

File: tools/edge_v/src/mesh/Refinement.cpp [hunk 5]
```diff
@@ -112,7 +112,7 @@ void edge_v::mesh::Refinement::init( std::size_t             i_nVes,
 #ifdef PP_USE_OMP
 }
 #endif
-  for( std::size_t l_ve = 0; l_ve < i_nVes; l_ve++ ) m_refVe[l_ve] = std::numeric_limits< double >::max();
+  for( std::size_t l_ve = 0; l_ve < i_nVes; l_ve++ ) m_refVe[l_ve] = std::numeric_limits< float >::max();
```
Comment (breuera) [inline tools/edge_v/src/mesh/Refinement.cpp]: Move this to a separate PR.

Comment (alheinecke) [reply tools/edge_v/src/mesh/Refinement.cpp]: can you just fix it manually?

File: src/data/MmXsmmFused.hpp (hunk 2 above)
Comment (alheinecke) [reply src/data/MmXsmmFused.hpp]: I used PP_N_CRUNS as this is used in the constructor. Should we also change the constructor then?

File: src/impl/seismic/kernels/SurfIntFused.hpp (hunk 4 above)
Comment (alheinecke) [reply src/impl/seismic/kernels/SurfIntFused.hpp]: it should be in this fine such that the compiler could inline. In fact a macro might be the best....

**Comment by alheinecke** [comment]
update which should address the issues pointed out

File: src/data/MmXsmmFused.hpp (hunk 2 above)
Comment (breuera) [reply src/data/MmXsmmFused.hpp]: No, guess it's fine for now.

File: src/impl/seismic/kernels/SurfIntFused.hpp (hunk 4 above)
Comment (breuera) [reply src/impl/seismic/kernels/SurfIntFused.hpp]: Ok; inline works.

File: tools/edge_v/src/mesh/Refinement.cpp (hunk 5 above)
Comment (breuera) [reply tools/edge_v/src/mesh/Refinement.cpp]: Yes, it's already fixed.

File: src/impl/seismic/kernels/SurfIntFused.hpp [hunk 6]
```diff
@@ -458,31 +497,31 @@ class edge::seismic::kernels::SurfIntFused: public edge::seismic::kernels::SurfI
...
                               o_scratch[1][0][0],
                               nullptr,
                               (l_fa < TL_N_FAS_DIV2) ? i_dofsP[0][0] : i_tDofsP[0][0],
                               nullptr );
 
         // transposed flux matrix
-        m_mm.m_kernels[2][l_fa]( o_scratch[1][0][0],
+        execMmKernel( 2, l_fa, o_scratch[1][0][0],
                                  m_fIntT[l_fa],
                                  io_dofsE[0][0] );
```
Comment (breuera) [inline src/impl/seismic/kernels/SurfIntFused.hpp]: Double check indentation throughout the PR.

File: src/impl/seismic/kernels/SurfIntFused.hpp [hunk 7]
```diff
@@ -382,6 +386,41 @@ class edge::seismic::kernels::SurfIntFused: public edge::seismic::kernels::SurfI
       }
     }
 
+    inline void execMmKernel( unsigned short   i_group, 
+                              unsigned short   i_kernel,
+                              const TL_T_REAL *i_a,
+                              const TL_T_REAL *i_b,
+                                    TL_T_REAL *io_c ) {
```
Comment (breuera) [inline src/impl/seismic/kernels/SurfIntFused.hpp]: Change to (consistency with the newer parts of the code):

```
inline void execMmKernel( unsigned short         i_group, 
                          unsigned short         i_kernel,
                          TL_T_REAL      const * i_a,
                          TL_T_REAL      const * i_b,
                          TL_T_REAL            * io_c ) {
```

**Comment by alheinecke** [comment]
I manually merged the full flexible N_CRUNS into this commit and fixed formatting
